go-controller/pkg/ovn/namespace.go

```diff
@@ -105,28 +105,53 @@ func (oc *Controller) multicastUpdateNamespace(ns *kapi.Namespace, nsInfo *names
 	}
 
 	var err error
+	nsInfo.multicastEnabled = enabled
 	if enabled {
 		err = oc.createMulticastAllowPolicy(ns.Name, nsInfo)
 	} else {
-		err = deleteMulticastAllowPolicy(ns.Name)
+		err = deleteMulticastAllowPolicy(ns.Name, nsInfo)
 	}
 	if err != nil {
 		klog.Errorf(err.Error())
 		return
 	}
-
-	nsInfo.multicastEnabled = enabled
 }
 
 // Cleans up the multicast policy for this namespace if multicast was
 // previously allowed.
 func (oc *Controller) multicastDeleteNamespace(ns *kapi.Namespace, nsInfo *namespaceInfo) {
 	if nsInfo.multicastEnabled {
-		if err := deleteMulticastAllowPolicy(ns.Name); err != nil {
+		nsInfo.multicastEnabled = false
+		if err := deleteMulticastAllowPolicy(ns.Name, nsInfo); err != nil {
 			klog.Errorf(err.Error())
 		}
 	}
-	nsInfo.multicastEnabled = false
+}
+
+// returns the UUID of a port_group for the given namespace if it does not exist create it
+func (nsInfo *namespaceInfo) ensureNamespacePortGroup(ns string) (string, string, error) {
+	// If the port_group already exists return it
+	if nsInfo.namespacePortGroupUUID != "" {
+		return nsInfo.namespacePortGroupUUID, hashedPortGroup(ns), nil
+	}
+
+	// Create the port_group if it does not exist
+	portGroupUUID, err := createPortGroup(ns, hashedPortGroup(ns))
+	if err != nil {
+		return "", "", fmt.Errorf("Failed to create port_group for %s (%v)",
+			ns, err)
+	}
+	nsInfo.namespacePortGroupUUID = portGroupUUID
+
+	return nsInfo.namespacePortGroupUUID, hashedPortGroup(ns), nil
+
+}
+
+// only delete the namespace port group if all referances to it are disabled
+func (nsInfo *namespaceInfo) deleteNamespacePortGroup(ns string) {
+	if !nsInfo.multicastEnabled {
```

Comment:
The name `deleteNamespacePortGroup` makes it sound like it *definitely* deletes it so it's weird that it doesn't.

I'm not sure how you're planning to change this with the egress firewall stuff... But maybe there should just be a single `updateNamespacePortGroup` which creates it if "it should be there but it isn't" and deletes it if "it shouldn't be there but it is"? The loadbalancer-syncing code works that way IIRC.

Reply:
changed it to cleanupNamespacePortGroup I think it looks cleaner to have the two different functions I will look into changing it

Reply:
changed this to updateNamespacePortGroup()